File: library/src/main/java/com/nshmura/strictmodenotifier/ViolationTypeInfo.java
package com.nshmura.strictmodenotifier;

import android.os.Build;
import com.nshmura.strictmodenotifier.detector.ClassInstanceLimitDetector;
import com.nshmura.strictmodenotifier.detector.CleartextNetworkDetector;
import com.nshmura.strictmodenotifier.detector.CustomSlowCallDetector;
import com.nshmura.strictmodenotifier.detector.Detector;
import com.nshmura.strictmodenotifier.detector.FileUriExposureDetector;
import com.nshmura.strictmodenotifier.detector.LeakedClosableObjectsDetector;
import com.nshmura.strictmodenotifier.detector.NetworkDetector;
import com.nshmura.strictmodenotifier.detector.ResourceMismatchDetector;

public enum ViolationTypeInfo {

  // ThreadPolicy
  CUSTOM_SLOW_CALL(
      "Custom Slow Call",
      ViolationType.CUSTOM_SLOW_CALL,
      Build.VERSION_CODES.HONEYCOMB,
      new CustomSlowCallDetector()),

  NETWORK(
      "Network",
      ViolationType.NETWORK,
      Build.VERSION_CODES.GINGERBREAD,
      new NetworkDetector()),

  RESOURCE_MISMATCHES(
      "Resource Mismatches",
      ViolationType.RESOURCE_MISMATCHES,
      Build.VERSION_CODES.M,
      new ResourceMismatchDetector()),

  // VmPolicy
  CLASS_INSTANCE_LIMIT(
      "Class Instance Limit",
      ViolationType.CLASS_INSTANCE_LIMIT,
      Build.VERSION_CODES.HONEYCOMB,
      new ClassInstanceLimitDetector()),

  CLEARTEXT_NETWORK(
      "Cleartext Network",
      ViolationType.CLEARTEXT_NETWORK,
      Build.VERSION_CODES.M,
      new CleartextNetworkDetector()),

  FILE_URI_EXPOSURE(
      "File Uri Exposure",
      ViolationType.FILE_URI_EXPOSURE,
      Build.VERSION_CODES.JELLY_BEAN_MR2,
      new FileUriExposureDetector()),

  LEAKED_CLOSABLE_OBJECTS(
      "Leaked Closable Objects",
      ViolationType.LEAKED_CLOSABLE_OBJECTS,
      Build.VERSION_CODES.HONEYCOMB,
      new LeakedClosableObjectsDetector()),

  ACTIVITY_LEAKS(
      "Activity Leaks",
      ViolationType.ACTIVITY_LEAKS,
      Build.VERSION_CODES.HONEYCOMB,
      null),

  LEAKED_REGISTRATION_OBJECTS(
      "Leaked Registration_Objects",
      ViolationType.LEAKED_REGISTRATION_OBJECTS,
      Build.VERSION_CODES.JELLY_BEAN,
      null),

  LEAKED_SQL_LITE_OBJECTS(
      "Leaked Sql Lite Objects",
      ViolationType.LEAKED_SQL_LITE_OBJECTS,
      Build.VERSION_CODES.GINGERBREAD,
      null),

  UNKNOWN("UNKNOWN", ViolationType.UNKNOWN, 0, null);

  private String name;
  public final ViolationType violationType;
  public final int minSdkVersion;
  public final Detector detector;

  ViolationTypeInfo(String name, ViolationType violationType, int minSdkVersion,
      Detector detector) {
    this.name = name;
    this.violationType = violationType;
    this.minSdkVersion = minSdkVersion;

    if (detector != null) {
      this.detector = detector;
    } else {
      this.detector = new Detector() {
        @Override public boolean detect(StrictModeLog log) {
          return false;
        }
      };
    }
  }

  public static ViolationTypeInfo convert(ViolationType type) {
    for (ViolationTypeInfo info : values()) {
      if (info.violationType == type) {
        return info;
      }
    }
    return ViolationTypeInfo.UNKNOWN;
  }

  public String violationName() {
    return name + " Violation";
  }
}

File: notifier/src/main/java/io/github/juanimoli/strictmode/notifier/ViolationTypeInfo.kt
package io.github.juanimoli.strictmode.notifier

import android.os.Build.VERSION_CODES
import io.github.juanimoli.strictmode.notifier.commons.ViolationType
import io.github.juanimoli.strictmode.notifier.detector.*

enum class ViolationTypeInfo(
    val violationName: String,
    val violationType: ViolationType,
    val minSdkVersion: Int,
    detector: Detector?
) {
    // ThreadPolicy
    CUSTOM_SLOW_CALL(
        "Custom Slow Call",
        ViolationType.CUSTOM_SLOW_CALL,
        VERSION_CODES.HONEYCOMB,
        CustomSlowCallDetector()
    ),
    NETWORK(
        "Network",
        ViolationType.NETWORK,
        VERSION_CODES.GINGERBREAD,
        NetworkDetector()
    ),
    RESOURCE_MISMATCHES(
        "Resource Mismatches",
        ViolationType.RESOURCE_MISMATCHES,
        VERSION_CODES.M,
        ResourceMismatchDetector()
    ),  // VmPolicy
    CLASS_INSTANCE_LIMIT(
        "Class Instance Limit",
        ViolationType.CLASS_INSTANCE_LIMIT,
        VERSION_CODES.HONEYCOMB,
        ClassInstanceLimitDetector()
    ),
    CLEARTEXT_NETWORK(
        "Cleartext Network",
        ViolationType.CLEARTEXT_NETWORK,
        VERSION_CODES.M,
        CleartextNetworkDetector()
    ),
    FILE_URI_EXPOSURE(
        "File Uri Exposure",
        ViolationType.FILE_URI_EXPOSURE,
        VERSION_CODES.JELLY_BEAN_MR2,
        FileUriExposureDetector()
    ),
    LEAKED_CLOSABLE_OBJECTS(
        "Leaked Closable Objects",
        ViolationType.LEAKED_CLOSABLE_OBJECTS,
        VERSION_CODES.HONEYCOMB,
        LeakedClosableObjectsDetector()
    ),
    ACTIVITY_LEAKS(
        "Activity Leaks",
        ViolationType.ACTIVITY_LEAKS,
        VERSION_CODES.HONEYCOMB,
        null
    ),
    LEAKED_REGISTRATION_OBJECTS(
        "Leaked Registration_Objects",
        ViolationType.LEAKED_REGISTRATION_OBJECTS,
        VERSION_CODES.JELLY_BEAN,
        null
    ),
    LEAKED_SQL_LITE_OBJECTS(
        "Leaked Sql Lite Objects",
        ViolationType.LEAKED_SQL_LITE_OBJECTS,
        VERSION_CODES.GINGERBREAD,
        null
    ),
    UNKNOWN("UNKNOWN", ViolationType.UNKNOWN, 0, null);

    var detector: Detector? = null

    fun violationName(): String {
        return "$violationName Violation"
    }

    companion object {
        fun convert(type: ViolationType?): ViolationTypeInfo {
            for (info in values()) {
                if (info.violationType == type) {
                    return info
                }
            }

            return UNKNOWN
        }
    }

    init {
        if (detector != null) {
            this.detector = detector
        } else {
            this.detector = object : Detector {
                override fun detect(log: StrictModeLog): Boolean {
                    return false
                }
            }
        }
    }
}